src/Schema/Directives/WithRelationDirective.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+namespace Nuwave\Lighthouse\Schema\Directives;
+
+use Closure;
+use GraphQL\Deferred;
+use GraphQL\Type\Definition\ResolveInfo;
+use Illuminate\Database\Eloquent\Model;
+use Nuwave\Lighthouse\Execution\DataLoader\BatchLoader;
+use Nuwave\Lighthouse\Schema\Values\FieldValue;
+use Nuwave\Lighthouse\Support\Contracts\GraphQLContext;
+
+abstract class WithRelationDirective extends BaseDirective
+{
+    /**
+     * The fully-qualified class name of the batch loader to use.
+     *
+     * @return class-string<\Nuwave\Lighthouse\Execution\DataLoader\BatchLoader>
+     */
+    abstract protected function batchLoaderClass(): string;
+
+    /**
+     * The name of the relation to be loaded.
+     */
+    abstract protected function relationName(): string;
+
+    /**
+     * Eager load a relation on the parent instance.
+     */
+    public function handleField(FieldValue $fieldValue, Closure $next): FieldValue
+    {
+        return $next(
+            $fieldValue->setResolver(
+                $this->deferredRelationResolver(
+                    $fieldValue->getResolver()
+                )
+            )
+        );
+    }
+
+    /**
+     * Decorate the builder used to fetch the models.
+     */
+    protected function decorateBuilder(ResolveInfo $resolveInfo): Closure
+    {
+        return function ($query) use ($resolveInfo) {
+            $resolveInfo->argumentSet->enhanceBuilder(
+                $query,
+                $this->directiveArgValue('scopes', [])
+            );
+        };
+    }
+
+    /**
+     * Return a new deferred resolver.
+     *
+     * @param  callable  $resolver
+     */
+    protected function deferredRelationResolver($resolver): Closure
```

Comment:
```suggestion
    protected function deferredRelationResolver(callable $resolver): Closure
```

Really like how you have broken this apart, good stuff.